### PR TITLE
Add 1d and RZ support to mewarpx

### PR DESCRIPTION
**Pull request by roelof-groenewald**
The `init_grid` function of `mwxrun.py` has been heavily refactored to make things general for all the different simulation dimensional cases. Furthermore, a `ZDiscEmitter` was added and the `VolumetricInjector` was modified to be able to handle both Cartesian and cylindrical coordinate systems.

One downside of these changes is that circleCI now has to build WarpX for 1d, 2d and RZ, which increases the build time on circleCI from ~10 minutes to ~20 minutes.

The current Poisson solver for RZ fails to converge for most of the vacuum diode simulations I tried to run. However, a fix for this is in the works:
- [x] https://github.com/ECP-WarpX/WarpX/pull/2602

With the changes that will be included in the above PR and using Weiqun's updated `MLEBNodeFDLaplacian` (from https://github.com/AMReX-Codes/amrex/pull/2538), I get the following results for a vacuum diode simulation (using `r_max` = 200 um and setting the cathode radius to 150 um):

D_50

Currently, Schottky enhancement is not supported in RZ since `compute_weights_nodal` (from `Source/Particles/Gather/ScalarFieldGather.H`) does not support RZ.

The inclusion of 1d support was tested with the Turner benchmark and gave expected results:

warpX_benchmark_Turner2013

### CODE REVIEW

File: mewarpx/mewarpx/setups_store/diode_setup.py [hunk 1]
```diff
@@ -167,24 +167,27 @@ class DiodeRun_V1(object):
     # steps between doing load-balancing. If <1 load balancing will not be done
     LOAD_BALANCE_INTERVALS = 0
 
-    def __init__(self, dim=1, rz=False, **kwargs):
+    def __init__(self, **kwargs):
         for kw in list(kwargs.keys()):
             setattr(self, kw, kwargs[kw])
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/setups_store/diode_setup.py]: Small suggestion that I should've made when this code was first written, but I didn't know this at the time. It's faster to iterate through all the items of the dictionary than to just iterate through the keys.
```suggestion
        for kw, value in kwargs.items():
            setattr(self, kw, value)
```

File: mewarpx/mewarpx/setups_store/diode_setup.py [hunk 2]
```diff
@@ -167,24 +167,27 @@ class DiodeRun_V1(object):
...
 
-        self.dim = dim
+        self.dim = mwxrun.dim
         if self.dim not in [1, 2, 3]:
             raise ValueError(f"Unavailable dimension {self.dim}")
-        self.rz = rz
+        self.rz = mwxrun.geom_str == 'RZ'
         if self.dim != 2 and self.rz:
             raise ValueError("dim=2 required for RZ")
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/setups_store/diode_setup.py:181]: Doing this forces us to always have to call `mwxutil.init_libwarpx` at the top of any input script, even if we find a solution for our import `_libwarpx` use case.  Maybe we can add logic here to say that `self.dim` and `self.rz` get set to their `mwxrun` counterparts if the `mwxrun` attributes are set. Otherwise, set the `mwxrun` attributes here based on `self.dim` and `self.rz`.

File: mewarpx/mewarpx/setups_store/diode_setup.py [hunk 3]
```diff
@@ -413,22 +407,25 @@ def init_inert_gas(self):
...
-                    required_precision=1e-6,
-                    maximum_iterations=10000
-                )
+        else:
+            self.solver = picmi.ElectrostaticSolver(
+                grid=mwxrun.grid,
+                method='Multigrid',
+                required_precision=1e-6,
+                maximum_iterations=10000
+            )
```
Comment (KZhu-ME) [inline mewarpx/mewarpx/setups_store/diode_setup.py:430]: Would it be useful to make another PR to allow `diode_setup` to set an absolute error for the electrostatic solver?

File: mewarpx/tests/test_mwxrun.py [hunk 4]
```diff
@@ -87,26 +78,6 @@ def test_capacitive_discharge_multigrid(caplog, name):
...
-    ds = yt.load( "diags/fields/fields00010" )
-    grid_data = ds.covering_grid(
-        level=0, left_edge=ds.domain_left_edge, dims=ds.domain_dimensions
-    )
-    rho_data_yt = np.mean(
-        grid_data['rho_he_ions'].to_ndarray()[:,:,0], axis=0
-    )[2:-2]
-
-    assert np.allclose(rho_data, rho_data_yt, rtol=0.01)
-
```
Comment (KZhu-ME) [inline mewarpx/tests/test_mwxrun.py:110]: It would be useful to have a test for rz, 1d, 2d, and 3d that analyzes diag output to catch any unintended physics changes.

**Review by KZhu-ME — CHANGES_REQUESTED** [review]
Looks good, I just have a few suggestions. Also, as an overall comment, this PR would be a major version change for mewarpx right? Since we're making backwards incompatible api changes (mostly `mwxrun.init_grid`).

File: mewarpx/mewarpx/setups_store/diode_setup.py (hunk 2 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/setups_store/diode_setup.py:181]: Based on the direction David went with the `libwarpx` refactor I don't see us having a way to avoid calling `mwxutil.init_libwarpx` anytime soon. I like the changes here in that it unifies specifying the simulation dimension to a single place and therefore avoids the possibility to have different parts of the code think different dimensions are used.

File: mewarpx/mewarpx/setups_store/diode_setup.py (hunk 3 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/setups_store/diode_setup.py:430]: Yeah absolutely, if you think there is a need for it go ahead.

File: mewarpx/tests/test_mwxrun.py (hunk 4 above)
Comment (roelof-groenewald) [reply mewarpx/tests/test_mwxrun.py:110]: We definitely should have tests like that, I agree. Not in this file though (which is just testing that execution works). I will add a physics test once the gating upstream PR is completed.

**Comment by roelof-groenewald** [comment]
@KZhu-ME Thanks for the review. Yes this will include a major version change. I'm waiting on this being ready to go before updating the changelog and version number since I assume there will be several other PRs merged into `memaster` before this is done (avoiding extra merge conflicts to sort out).

File: mewarpx/mewarpx/diags_store/field_diagnostic.py [hunk 5]
```diff
@@ -118,6 +118,9 @@ def fields_diag(self):
 
         if self.process_phi:
             data = mwxrun.get_gathered_phi_grid(include_ghosts=False)
+            if mwxrun.dim == 1:
+                data = np.tile(data, mwxrun.nz//2).reshape(
+                    mwxrun.nz//2, data.shape[0])
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/diags_store/field_diagnostic.py]: Is there a comment about why the `//2` is here?

File: mewarpx/mewarpx/emission.py [hunk 6]
```diff
@@ -1186,6 +1182,108 @@ def get_normals(self, x, y, z):
...
+    """This injects over an x-y disc rather than a rectangle."""
+    geoms = ['RZ']
+
+    def __init__(self, conductor, T, inner_emission_radius=None,
+                 outer_emission_radius=None, transverse_fac=1.0, **kwargs):
+        """Initialize an emitter for a disc (circular) cathode.
+
+        Arguments:
+            conductor (warp.Assembly): Conductor object, used to obtain work
+                function. Can later grab other variables from this conductor.
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]:
```suggestion
                function and z coordinate/direction.
```
Also a similar edit to `ZPlaneEmitter`

File: mewarpx/mewarpx/emission.py [hunk 7]
```diff
@@ -1186,6 +1182,108 @@ def get_normals(self, x, y, z):
...
+            T (float): Temperature in Kelvin for the emitter; determines
+                velocities.
+            inner_emission_radius (float): Inner radius of the disc (in meters)
+                for particles to be emitted from. Default mwxrun.rmin.
+            outer_emission_radius (float): Outer radius of the disc (in meters)
+                for particles to be emitted from. Default mwxrun.rmax.
+            transverse_fac (float): Scale the transverse energy distribution by
+                this factor. Default 1. See
+                :func:`metools.runtools.get_velocities` for details.
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: Stale reference.

File: mewarpx/mewarpx/emission.py [hunk 8]
```diff
@@ -1186,6 +1182,108 @@ def get_normals(self, x, y, z):
...
+            kwargs (dict): Any other keyword arguments supported by the parent
+                Emitter constructor (such as "use_Schottky" or
+                "emission_type").
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: Is use_Schottky also outdated here?

File: mewarpx/mewarpx/emission.py [hunk 9]
```diff
@@ -1186,6 +1182,108 @@ def get_normals(self, x, y, z):
...
+        if inner_emission_radius is None:
+            inner_emission_radius = mwxrun.rmin
+        self.inner_emission_radius = inner_emission_radius
+        if outer_emission_radius is None:
+            outer_emission_radius = mwxrun.rmax
+        self.outer_emission_radius = outer_emission_radius
+
+        if mwxrun.rmin > 0:
+            raise RuntimeError("r_min > 0 not yet implemented for ZDiscEmitter")
+
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: Why is this? I don't see a problem in this class specifically.

File: mewarpx/mewarpx/emission.py [hunk 10]
```diff
@@ -1186,6 +1182,108 @@ def get_normals(self, x, y, z):
...
+            x (np.ndarray): x-coordinates of emitted particles (in meters).
+            y (np.ndarray): y-coordinates of emitted particles (in meters).
+            z (np.ndarray): z-coordinates of emitted particles (in meters).
+
+        Returns:
+            normals (np.ndarray): nx3 array containing the outward surface
+                normal vector at each particle location.
+        """
+        normals = np.zeros((len(x), 3))
+        normals[:, 2] = 1.0
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]:
```suggestion
        normals[:, 2] = -self.zsign
```

File: mewarpx/mewarpx/emission.py [hunk 11]
```diff
@@ -1461,64 +1559,90 @@ class VolumeEmitter(BaseEmitter):
...
-    def getvoltage(self):
-        """Ideally this is probably the local potential, but default to 0."""
-        return 0.
-
-    def getvoltage_e(self):
-        """Ideally this is probably the local potential, but default to 0."""
-        return self.getvoltage()
-
+        # determine default prism type from simulation geometry
+        self.rectangular = mwxrun.geom_str != 'RZ'
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py:1580]: (1) This is now redundant with setting it in lines 1601-1604 I believe.
(2) If sticking with a bool, I would prefer a `self.cylindrical` bool rather than `self.rectangular`, because rectangular feels more like the default case. Something not being rectangular doesn't communicate to me that it's cylindrical.

File: mewarpx/mewarpx/emission.py [hunk 12]
```diff
@@ -1461,64 +1559,90 @@ class VolumeEmitter(BaseEmitter):
...
+        )
 
-    A template ``_get_xv_coords()`` is also provided but can be overridden if
-    needed. It calls ``_get_x_coords()``.
-    """
+        if r_bounds_given and xy_bounds_given:
+            raise AttributeError(
+                "Both rectangular and cylindrical boundaries specified for a "
+                "VolumeEmitter"
+            )
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py:1593]: Also add a check if neither are given.

File: mewarpx/mewarpx/emission.py [hunk 13]
```diff
@@ -1461,64 +1559,90 @@ class VolumeEmitter(BaseEmitter):
...
+        if r_bounds_given:
+            self.rectangular = False
+        if xy_bounds_given:
+            self.rectangular = True
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py:1597]:
```suggestion
        # We know r_bounds_given T/F => xy_bounds_given F/T
        self.cylindrical = r_bounds_given
```

File: mewarpx/mewarpx/emission.py [hunk 14]
```diff
@@ -1461,64 +1559,90 @@ class VolumeEmitter(BaseEmitter):
...
-                 mwxrun.ymax, mwxrun.zmin, mwxrun.zmax])
-        ):
-            if lim is None:
-                lim = defaultlim
-            self.bounds[ii // 2, ii % 2] = lim
+        else:
+            for ii, (lim, defaultlim) in enumerate(
+                zip([rmin, rmax, 0, 2.0*np.pi, zmin, zmax],
+                    [mwxrun.xmin, mwxrun.xmax, mwxrun.ymin,
+                     mwxrun.ymax, mwxrun.zmin, mwxrun.zmax])
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/emission.py]: Need to change defaults I think.

File: mewarpx/mewarpx/mwxrun.py [hunk 15]
```diff
@@ -53,57 +53,86 @@ class MEWarpXRun(object):
...
-            upper_boundary_conditions_particles=upper_boundary_conditions_parts,
-            warpx_max_grid_size=nz//min_tiles
+        # sanity check inputs
+        assert len(lower_bound) == self.dim
+        assert len(upper_bound) == self.dim
+        assert len(number_of_cells) == self.dim
+
+        boundary_conditions = {'fields': {}, 'particles': {}}
+
+        # set boundary conditions to be periodic in x and y except for RZ
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/mwxrun.py]:
```suggestion
        # set default boundary conditions to be periodic in x and y except for RZ
```

File: mewarpx/mewarpx/mwxrun.py [hunk 16]
```diff
@@ -53,57 +53,86 @@ class MEWarpXRun(object):
...
-            upper_boundary_conditions=upper_boundary_conditions,
-            lower_boundary_conditions_particles=lower_boundary_conditions_parts,
-            upper_boundary_conditions_particles=upper_boundary_conditions_parts,
-            warpx_max_grid_size=nz//min_tiles
+        # sanity check inputs
+        assert len(lower_bound) == self.dim
+        assert len(upper_bound) == self.dim
+        assert len(number_of_cells) == self.dim
+
+        boundary_conditions = {'fields': {}, 'particles': {}}
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/mwxrun.py:85]:
```suggestion
        # Dictionary to first provide defaults, then overwrite with user settings.
        boundary_conditions = {'fields': {}, 'particles': {}}
```

File: mewarpx/mewarpx/mwxrun.py [hunk 17]
```diff
@@ -53,57 +53,86 @@ class MEWarpXRun(object):
...
+            boundary_conditions[kind][side][self.coord_map[dim]] = kwargs[key]
+
+        if self.geom_str == 'RZ':
+            grid = picmi.CylindricalGrid
+        elif self.dim == 1:
+            grid = picmi.Cartesian1DGrid
+        elif self.dim == 2:
+            grid = picmi.Cartesian2DGrid
+        elif self.dim == 3:
+            grid = picmi.Cartesian3DGrid
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/mwxrun.py]: I would check geom_str explicitly for Z/XZ/XYZ, then error if it's different, in case we later add a new geometry.

File: mewarpx/mewarpx/mwxrun.py [hunk 18]
```diff
@@ -298,6 +360,22 @@ def _set_grid_params(self):
         else:
             raise ValueError("Unrecognized type of pywarpx.picmi Grid.")
 
+    def _print_grid_params(self):
+        """Function to print out the grid parameters."""
+        if self.geom_str == 'RZ':
+            logger.info(
+                f"Creating grid with nr={self.nr}, ny={self.ny}, nz={self.nz} "
```
Comment (peterscherpelz) [inline mewarpx/mewarpx/mwxrun.py]: Is ny defined?

**Review by peterscherpelz — CHANGES_REQUESTED** [review]
I'll have to find time later to review further, but here's a start.

On the whole I don't really like the VolumeEmitter set the right rmin/rmax etc.; if nothing else it prevents default initialization without passing any arguments. I can't think of a reason to do cylindrical VolumeEmitter in Cartesian geometry; if that's the case I would vote for just matching simulation choice, and then erroring if incorrect bounds were passed whereas use proper defaults if not.
Or even if you do support cylindrical in a Cartesian grid, make that override the default, eg have a `cylindrical=None` parameter; If T/F then override the simulation default.

File: mewarpx/mewarpx/diags_store/field_diagnostic.py (hunk 5 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/diags_store/field_diagnostic.py]: It is completely arbitrary just to give the 2d plot a finite width, so not physical at all.

File: mewarpx/mewarpx/emission.py (hunk 12 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py:1593]: If neither are given the simulation domain will be used along with the simulation geometry. This also addresses the overall comment you made (defaults will be properly set using the simulation properties) and the comment above (the `rectangular` flag will only be overwritten if either rectangular or cylindrical boundaries are specified).

File: mewarpx/mewarpx/emission.py (hunk 14 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py]: Seeing as in `mwxrun` we set `self.xmin = self.ymin = self.rmin = self.grid.rmin` and same for `xmax` it is will work fine as is but happy to change it to make things clearer.

File: mewarpx/mewarpx/mwxrun.py (hunk 18 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/mwxrun.py]: Yeah, similarly as I commented above: in `mwxrun.py` we have `self.nx = self.ny = self.nr = self.grid.nr`

File: mewarpx/mewarpx/emission.py (hunk 9 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py]: That is outdated. Thanks for catching!

File: mewarpx/mewarpx/emission.py (hunk 8 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py]: Indeed yes. Thanks for catching!

File: mewarpx/mewarpx/emission.py (hunk 11 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py:1580]: I prefer `rectangular` because rectangular feels more like the default. If we make it `self.cylindrical` then in the child classes `UniformDistributionVolumeEmitter` and `ZSinDistributionVolumeEmitter` we would have statements of the form `if not self.cylindrical` instead of `if self.rectangular`.

File: mewarpx/mewarpx/emission.py (hunk 12 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/emission.py:1593]: I do agree that the cylindrical prism won't really be useful with a Cartesian grid until we consider 3d simulations but I don't see a reason not to build the code as general as possible from the start.

File: mewarpx/mewarpx/diags_store/field_diagnostic.py (hunk 5 above)
Comment (roelof-groenewald) [reply mewarpx/mewarpx/diags_store/field_diagnostic.py]: I changed this to only tile the data being plotted but not the array that is saved.

**Review by peterscherpelz — APPROVED** [review]
Thanks! No further comments. Sorry for completely messing up the rectangular volume emitter min/max etc. logic on my first review...

**Comment by roelof-groenewald** [comment]
> Thanks! No further comments. Sorry for completely messing up the rectangular volume emitter min/max etc. logic on my first review...

Thanks and no worries. It is easy to clear up misunderstandings so any questions/concerns that come up are always welcome.

**Review by KZhu-ME — APPROVED** [review]
Looks good!
